### ISSUE TITLE
Hanamonogatari incorrect partial map

### ISSUE THREAD

**Issue by echen92** [ZeroQI/Hama.bundle#490]
Clean library with only one folder for the Hanamonogatari arc.
Matches the show name + description good but episodes come back with meta from bakemonogatari season 1.
Agent-update log seems to indicate it is mapping to both hanamono and bakemono in tvdb. Not using forced guid_type. 
s1e1_mapped shouldn't be true since the anime-list mapping from scudlee is AniDB S1 to TVDB S0?
I tried absolute numbering e.g s0E10 with anidb3 and episode meta mapped correctly but the show title + episode numbering pulled from TVDB.
Have tried the troubleshooting steps delete library->stopping server->wipe/update hama/ASS->restart&recreate library.

![screenshot](https://user-images.githubusercontent.com/1142325/144694993-ca4b174b-33f5-4f05-867a-d89435568249.jpg)

[com.plexapp.agents.hama.log](https://github.com/ZeroQI/Hama.bundle/files/7653213/com.plexapp.agents.hama.log)
[com.plexapp.system.log](https://github.com/ZeroQI/Hama.bundle/files/7653214/com.plexapp.system.log)
[Hanamonogatari.agent-search.log](https://github.com/ZeroQI/Hama.bundle/files/7653215/Hanamonogatari.agent-search.log)
[Hanamonogatari.agent-update.log](https://github.com/ZeroQI/Hama.bundle/files/7653226/Hanamonogatari.agent-update.log)

[Hanamonogatari.filelist.log](https://github.com/ZeroQI/Hama.bundle/files/7653216/Hanamonogatari.filelist.log)
[Hanamonogatari.scanner.log](https://github.com/ZeroQI/Hama.bundle/files/7653217/Hanamonogatari.scanner.log)
[_root_.agent.log](https://github.com/ZeroQI/Hama.bundle/files/7653218/_root_.agent.log)
[_root_.scanner.log](https://github.com/ZeroQI/Hama.bundle/files/7653219/_root_.scanner.log)

**Comment by ZeroQI** [comment]
Hanamonogatari ep 1-5 seen by scanner as season 1 ep 1-5 and matched to anidbid 10046 [https://anidb.net/anime/10046] 
You labelled incorrectly as AniDB 5 parts are showing as episodes 02-06 for AniDB. 

https://github.com/Anime-Lists/anime-lists/blob/master/anime-list-master.xml
```
 <anime anidbid="10046" tvdbid="102261" defaulttvdbseason="0" episodeoffset="" tmdbid="" imdbid="">
    <name>Hanamonogatari: Suruga Devil</name>
    <mapping-list>
      <mapping anidbseason="1" tvdbseason="0">;1-0;2-10;3-11;4-12;5-13;6-14;</mapping>
    </mapping-list>
  </anime>
```
This translates to:
- anidb s01e01=>tvdb s00e00 [inexistant]
- anidb s01e02=>tvdb s00e10
- anidb s01e03=>tvdb s00e11
- anidb s01e04=>tvdb s00e12
- anidb s01e05=>tvdb s00e13
- anidb s01e06=>tvdb s00e14

TVDB s00e10-13 are detected correctly on disk

> [X]   s0e10 s01e002 anidbid:   10046 air_date: 2014-08-16 abs_number: 0, title: Suruga Devil - Part 1
>  - [1] title:   [en] Suruga Devil - Part 1
>  - [1] summary: [en] After the graduation of Araragi and Senjougahara, Suruga alone to deal with her left hand that has been turned monsterous from her mother's monkey paw charm. After hearing rumors of a "Devil" that can solve one's problems no matter what, Suruga set out to find this person.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_4653123_ja.json'
>  - [ ] thumb: https://thetvdb.plexapp.com/banners/episodes/102261/4653123.jpg
> [X]   s0e11 s01e003 anidbid:   10046 air_date: 2014-08-16 abs_number: 0, title: Suruga Devil - Part 2
>  - [1] title:   [en] Suruga Devil - Part 2
>  - [1] summary: [en] After having her hand healed, Suruga has many questions for her former classmate Rouka. She was unable to contact her, however a meeting with Keiki Deshu, who seems to have survived the attack from the previous season, answers some questions of hers.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_4834515_ja.json'
>  - [ ] thumb: https://thetvdb.plexapp.com/banners/episodes/102261/4834515.jpg
> [X]   s0e12 s01e004 anidbid:   10046 air_date: 2014-08-16 abs_number: 0, title: Suruga Devil - Part 3
>  - [1] title:   [en] Suruga Devil - Part 3
>  - [1] summary: [en] Suruga realizes that the collector is in fact Rouka, and she has been collecting parts of the "devil" as she feeds off the misfortune of others. Her motives are questionable...
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_4834516_ja.json'
>  - [ ] thumb: https://thetvdb.plexapp.com/banners/episodes/102261/4834516.jpg
> [X]   s0e13 s01e005 anidbid:   10046 air_date: 2014-08-16 abs_number: 0, title: Suruga Devil - Part 4
>  - [1] title:   [en] Suruga Devil - Part 4
>  - [1] summary: [en] Rouka challenges Suruga to a one-on-one basketball match as the two discuss the events that led up to their collection of their respective "devil" parts. Suruga is still coming to terms with the ability to use her left arm again.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_4834517_ja.json'

TVDB s00e143 is not detected correctly

> [ ]   s0e14 s01e006 anidbid:   10046 air_date: 2014-08-16

TVDB S01e01-05 are detected wrongly and overwrite the metadata from s00exx

> [X]    s1e1 s01e001 anidbid: xxxxxxx air_date: 2009-07-03 abs_number: 1, title: Hitagi Crab - Part 1
>  - [1] title:   [en] Hitagi Crab - Part 1
>  - [1] summary: [en] After Koyomi Araragi catches his classmate, Hitagi Senjougahara, when she fell down some stairs, he makes a strange discovery about her. Trying to keep her secret Hitagi attacks Koyomi. However, he has a secret of his own and might be able to help her.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_830421_ja.json'
>  - [ ] thumb: https://thetvdb.plexapp.com/banners/episodes/102261/830421.jpg
> [X]    s1e2 s01e002 anidbid: xxxxxxx air_date: 2009-07-10 abs_number: 2, title: Hitagi Crab - Part 2
>  - [1] title:   [en] Hitagi Crab - Part 2
>  - [1] summary: [en] To overcome her problems, Hitagi has to face some difficult and painful memories from her past.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_830451_ja.json'
>  - [ ] thumb: https://thetvdb.plexapp.com/banners/episodes/102261/830451.jpg
> [X]    s1e3 s01e003 anidbid: xxxxxxx air_date: 2009-07-17 abs_number: 3, title: Mayoi Snail - Part 1
>  - [1] title:   [en] Mayoi Snail - Part 1
>  - [1] summary: [en] Koyomi Araragi has an argument with his sister and then finds it difficult to stay in the house. He goes to the park where he meets Hitagi Senjyogahara. They see a little girl, Mayoi Hachikuji, comparing the town map displayed in the park to the memo in her hand. Learning that Mayoi has lost her way and route to a relative's home, Koyomi offers to help, only to be violently rebuffed.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_882021_ja.json'
>  - [ ] thumb: https://thetvdb.plexapp.com/banners/episodes/102261/882021.jpg
> [X]    s1e4 s01e004 anidbid: xxxxxxx air_date: 2009-07-24 abs_number: 4, title: Mayoi Snail - Part 2
>  - [1] title:   [en] Mayoi Snail - Part 2
>  - [1] summary: [en] After Koyomi Araragi and Hitagi Senjyogahara discover that the girl they met in the park, Mayoi Hachikuji, is trying to get to her mother's house, they offer to walk along with her. However, they keep getting incredibly lost.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_908301_ja.json'
>  - [ ] thumb: https://thetvdb.plexapp.com/banners/episodes/102261/908301.jpg
> [X]    s1e5 s01e005 anidbid: xxxxxxx air_date: 2009-07-31 abs_number: 5, title: Mayoi Snail - Part 3
>  - [1] title:   [en] Mayoi Snail - Part 3
>  - [1] summary: [en] While Hitagi is off to Oshino for advice, Koyomi learns more about Mayoi's life. When Hitagi finally returns, she reveals some information that surprises Koyomi.
> common.SaveFile() - CachePath: '/volume1/Plex/Library/Application Support/Plex Media Server/Plug-in Support/Data/com.plexapp.agents.hama/DataItems', file: 'TheTVDB/json/102261/episode_935411_ja.json'
> 

We  should disregard new maping if earlier season mapping exist as failsafe
We should figure out why tvdb to AniDB mapping isn't working as expected

I am not in a position to reproduce the issue at the moment. Will update when i can

**Comment by Sn0wCrack** [comment]
I ran into this same issue with Nekomonogatari (Kuro), and to be fair the way the Monogatari Series get mapped to TVDB is kind of insane with a lot of the series mapping into random episodes of the Specials.

I attempted to try and fix this myself and I've got what seems to work, but unsure on its stability, I simply added the following to TheTVDBv2.py after it performs the episode translation:

```
      # If this episode already exists in our array it means we've already mapped it, so skip the rest
      # This can occur when something from say Season 0 or a Special is mapped into Season 1 by AnimeLists
      if Dict(TheTVDB_dict, 'seasons', season, 'episodes', episode, default=None):
        continue
```

Basically there didn't really appear to be anything checking if an episode already got mapped or not, so I just added a check for if it existed in the output dictionary, seems to work for this case at least, but doesn't really fix up how messed up Posters and such are for most of the Monogatari series entries given it all maps back to TVDB's Monogatari metaseries

**Comment by Sn0wCrack** [comment]
I've done some more testing with the above and ran into the opposite issue of Episodes mapping from Season 1 into Specials, ironically enough with Bakemonogatari, where Season 1 E13 - E15 get mapped into Season 0 E01 - E03.

I ended up making the following modification to track if an episode was written from a previous one and use that to skip processing the episode instead:

```python
      anidbid=""
      episode_moved_ahead = False
      abs_number = Dict(episode_json, 'absoluteNumber', default=0 if season=='0' else abs_number+1)
      if anidb_numbering:
        
        if Dict(mappingList, 'defaulttvdbseason_a'):
          new_season, new_episode = '1', str(abs_number)
        else: 
          new_season, new_episode, anidbid = AnimeLists.anidb_ep(mappingList, season, episode)
        
      elif metadata_source in ('tvdb3', 'tvdb4'):  
        for s in media.seasons:  #if abs id exists on disk, leave specials with no translation
          if str(abs_number) in list_abs_eps and str(abs_number) in media.seasons[s].episodes and s != "0": 
            new_season, new_episode = s, str(abs_number)
            break

      elif metadata_source=='tvdb5':  
        if abs_number:
          new_season, new_episode = '1', str(abs_number)

      episode_moved_ahead = (new_season > season or (new_season == season and new_episode > episode))

      episode = new_episode
      season = new_season
      
      # If this episode has been written to by a previous entry, such as Season S00E01 -> S010E01
      # we should skip over this one as it's been mapped into previously.
      if Dict(TheTVDB_dict, 'seasons', season, 'episodes', episode, 'episode_moved_ahead', default=False):
        continue

      SaveDict(episode_moved_ahead, TheTVDB_dict, 'seasons', season, 'episodes', episode, 'episode_moved_ahead')
```

This _should_ handle most of the cases I can think of, even should handle mapping just a single episode ahead.

Will probably test this myself for a bit and see if I run into any issues. If anyone knows of any other shows with funky mapping I'll see if I can test on those as well.

**Comment by ZeroQI** [comment]
Monogatari Series is the most complex series to organise, either anidb numbering and all in a collection or tvdb and potential mapping issues...

The issue with thumbnail, you cannot edit the priority afterwards, so it can give the illusion the default thumbnail is wrong, but i understand the full plex dance is annoying....

we need both a good failsafe, and review the mapping so it works without current known issues.
One of the other bugs brought a change that fixed a problem but brought mapping issues so we need to fix both at once
https://github.com/ZeroQI/Hama.bundle/issues/500
https://github.com/ZeroQI/Hama.bundle/issues/496

**Comment by Sn0wCrack** [comment]
Yeah, I quickly realised the Thumbnails were still there, just not set as the default "Poster."

I can take a look at some point to see if my changes in any way resolve #496 or #500 sometime in the next week or so.

**Comment by ZeroQI** [comment]
duplicate of
- https://github.com/ZeroQI/Hama.bundle/issues/500
- https://github.com/ZeroQI/Hama.bundle/issues/504

Will keep 500 which has better description and pinpointed the change that broke it
